### PR TITLE
Add initial KubeVirt API lifecycle automation and basic function test

### PR DESCRIPTION
**Pull request by Barakmor1**
This PR introduces the initial production code for the KubeVirt API lifecycle automation image, which automates machine type transitions to prevent upgrade failures when VMs run on obsolete machine types after nodes are upgraded to newer RHEL versions.

The following environment variables are exposed by the job:

- MACHINE_TYPE_GLOB (required): Regex to match machine types (e.g., \*rhel7\* targets RHEL 7 types).
- NAMESPACE (optional): Target namespace (defaults to all namespaces).
- RESTART_REQUIRED (optional): Boolean to restart VMs after patch (defaults to false).
- LABEL_SELECTOR (optional): Filter VMs by label (defaults to all VMs). Example: label1 in (value1,value2).

The job patches the VMs' machine types, and updates are applied after the next VM restart.

Additionally, this PR includes the first functional test for this automation.

Note: The skeleton CI for manifest generation, client generation, functional tests, unit tests, deployment of CI clusters, and builder logic is not included in this PR.

**Related links:**

This logic is inspired by @fossedihelm [work](https://github.com/kubevirt/kubevirt/commit/17923beed623998cc4216ef613dfb5e30732c647#diff-af5482e2e7fcc68f4899c67f7517a17c39609bb1498ba7816aa48f41db5136dbR105) with some changes.

Related [design proposal](https://github.com/kubevirt/community/pull/225) in kubevirt.

### CODE REVIEW

**Comment by Barakmor1** [comment]
/cc @acardace @xpivarc @enp0s3 @dasionov @kbidarkar

@acardace, @xpivarc, @fossedihelm 
, @enp0s3, @dasionov – You’ve been added as approvers and reviewers. 
The `approve` and `lgtm` labels aren’t available yet. Let me know if you’d prefer not to be included.

**Comment by enp0s3** [comment]
Hi, I will work on integrating the Openshift Prow bots for the merge automation

**Comment by fossedihelm** [comment]
I haven't gone into detail yet but at a glance I think files like `go.mod` or `Makefile` are missing.

**Comment by Barakmor1** [comment]
> I haven't gone into detail yet but at a glance I think files like `go.mod` or `Makefile` are missing.

Hey, you're right. I mentioned in the PR description that the automation and CI files are not included. They are already present here: https://github.com/openshift-virtualization/kubevirt-api-lifecycle-automation.

I focused this PR on the production code since it's the most critical part and needs to maintain backward compatibility. After we agree on this code I plan to open a follow-up PR for the automation and CI code, which can be reviewed separately. You're welcome to open PRs or review the existing files at any time.

I believe focusing on the production code now will result in higher-quality reviews, as it's easier to catch details with a smaller scope. However, if you'd prefer a single PR with everything included, I can close this one and submit a new PR with all the code.

@acardace @enp0s3 @fossedihelm @xpivarc @dasionov
Please let me know your preference.

If we stick with this PR, you can check out the commits locally to test and try them out with `make cluster-up && make cluster-sync && make functest` like with `kubevirt/kubevirt`

File: pkg/client-go/kubevirt/clientset/versioned/clientset.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2023 The KubevirtApiLifecycleAutomation Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (dasionov) [inline pkg/client-go/kubevirt/clientset/versioned/clientset.go]: i know it's a generated code, but we should probably update the year to 2024, and also I didn't see any licence so do we even need that ?

File: pkg/kubevirt-api-lifecycle-automation/application.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+const (
+	machineTypeEnvName     = "MACHINE_TYPE_GLOB"
+	namespaceEnvName       = "NAMESPACE"
+	restartRequiredEnvName = "RESTART_REQUIRED"
+	labelSelectorEnvName   = "LABEL_SELECTOR"
+)
+
+type MachineTypeUpdater struct {
+	virtClient      client.KubevirtApiLifecycleAutomationClient
+	machineTypeGlob string
```
Comment (dasionov) [inline pkg/kubevirt-api-lifecycle-automation/application.go:50]: this `machineTypeGlob` variable name is a bit confusing.. can i suggest you leave a comment for an example of the value it shell contain, if it's a single value like `pc-q35-rhel10.4.0` or a regex value?

File: pkg/kubevirt-api-lifecycle-automation/application.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+func (c *MachineTypeUpdater) patchMachineType(vm *virtv1.VirtualMachine) error {
+	// removing the machine type field from the VM spec reverts it to
+	// the default machine type of the VM's arch
+	patches := []patch.PatchOperation{
+		{
+			Op:    patch.PatchTestOp,
+			Path:  "/spec/template/spec/domain/machine/type",
+			Value: vm.Spec.Template.Spec.Domain.Machine.Type,
+		},
```
Comment (dasionov) [inline pkg/kubevirt-api-lifecycle-automation/application.go]: why do we need this?

File: pkg/kubevirt-api-lifecycle-automation/resources/operator/operator.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	return cj
+}
+
+func boolPtr(b bool) *bool {
+	return &b
+}
+
+func int64Ptr(i int64) *int64 {
+	return &i
+}
```
Comment (dasionov) [inline pkg/kubevirt-api-lifecycle-automation/resources/operator/operator.go]: do we still need those, or we can switch the the `pointer` package?

File: pkg/kubevirt-api-lifecycle-automation/application.go (hunk 2 above)
Comment (Barakmor1) [reply pkg/kubevirt-api-lifecycle-automation/application.go:50]: I plan to add the missing example to the official docs, probably in a follow-up PR. However, I’ve already included an example in the PR description and commit message. 

I’m open to suggestions for better names, naming can be hard.

File: pkg/client-go/kubevirt/clientset/versioned/clientset.go (hunk 1 above)
Comment (Barakmor1) [reply pkg/client-go/kubevirt/clientset/versioned/clientset.go]: I think we can modify it to 2025 at this point :)
Done

File: pkg/kubevirt-api-lifecycle-automation/application.go (hunk 3 above)
Comment (Barakmor1) [reply pkg/kubevirt-api-lifecycle-automation/application.go]: Removed, thanks!

File: pkg/kubevirt-api-lifecycle-automation/resources/operator/operator.go (hunk 4 above)
Comment (Barakmor1) [reply pkg/kubevirt-api-lifecycle-automation/resources/operator/operator.go]: Done

File: pkg/kubevirt-api-lifecycle-automation/application.go (hunk 2 above)
Comment (dasionov) [reply pkg/kubevirt-api-lifecycle-automation/application.go:50]: Turns out glob pattern is a thing i never knew at the time i wrote this comment, so for now i think it's suffice.
If you want to be extra readable, rename it to `machineTypeGlobPattern`.
But i don't think it's that important now

**Comment by dasionov** [comment]
/lgtm
Thank You!

File: pkg/kubevirt-api-lifecycle-automation/application.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+	virtCli, err := client.GetKubevirtApiLifecycleAutomationClient()
+	if err != nil {
+		log.Log.Errorf("Error retrieving virt client: %v", err)
+		os.Exit(1)
+	}
+
+	app, err := NewMachineTypeUpdater(virtCli)
+	if err != nil {
+		os.Exit(1)
```
Comment (enp0s3) [inline pkg/kubevirt-api-lifecycle-automation/application.go:70]: Missing error print here `log.Log.Errorf`

File: pkg/kubevirt-api-lifecycle-automation/application.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+	namespacesList, err := c.virtClient.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	namespaces := make([]string, len(namespacesList.Items))
+	for i, ns := range namespacesList.Items {
+		namespaces[i] = ns.Name
+	}
```
Comment (enp0s3) [inline pkg/kubevirt-api-lifecycle-automation/application.go]: Is this more efficient than `append`?

File: pkg/kubevirt-api-lifecycle-automation/application.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+		return nil
+	}
+
+	if runStrategy == virtv1.RunStrategyOnce {
+		return c.virtClient.KubevirtClient().KubevirtV1().VirtualMachines(vm.Namespace).Stop(context.Background(), vm.Name, &virtv1.StopOptions{})
+
+	}
+
+	return c.virtClient.KubevirtClient().KubevirtV1().VirtualMachines(vm.Namespace).Restart(context.Background(), vm.Name, &virtv1.RestartOptions{})
+}
```
Comment (enp0s3) [inline pkg/kubevirt-api-lifecycle-automation/application.go:164]: IIUC This should be conditional based on `RESTART_REQUIRED`

File: pkg/kubevirt-api-lifecycle-automation/application.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	patches := []patch.PatchOperation{
+		{
+			Op:   patch.PatchRemoveOp,
+			Path: "/spec/template/spec/domain/machine",
+		},
+	}
+
+	payload, err := patch.GeneratePatchPayload(patches...)
+	if err != nil {
+		// This is a programmer's error and should not happen
```
Comment (enp0s3) [inline pkg/kubevirt-api-lifecycle-automation/application.go]: Comment is redundant IMO

File: pkg/kubevirt-api-lifecycle-automation/application_test.go [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+			_, err := NewMachineTypeUpdater(kubevirtApiLifecycleAutomationClient)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring(fmt.Sprintf("error parsing %s environment variable, value \"%s\"", restartRequiredEnvName, badBoolean)))
+		})
+
+		When("it is true", func() {
+			BeforeEach(func() {
+				var err error
+				err = EnvVarManager.Setenv(machineTypeEnvName, machineTypeGlob)
+				Expect(err).ToNot(HaveOccurred())
```
Comment (enp0s3) [inline pkg/kubevirt-api-lifecycle-automation/application_test.go:242]: Missing test case when restart required is false

**Review by enp0s3 — COMMENTED** [review]
@Barakmor1 Thank you, that's a significant effort. So far reviewed the 1st commit, I have a few comments below

File: pkg/kubevirt-api-lifecycle-automation/resources/operator/operator.go (hunk 4 above)
Comment (enp0s3) [reply pkg/kubevirt-api-lifecycle-automation/resources/operator/operator.go]: @Barakmor1 Did you decide to switch?

Comment (Barakmor1) [reply pkg/kubevirt-api-lifecycle-automation/resources/operator/operator.go]: Yes on the last commit

File: tests/kubevirt_api_lifecycle_automation_test.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+	Context("Kubevirt Api Lifecycle Automation Job", func() {
+		BeforeEach(func() {
+			once.Do(func() {
+				c, err := f.GetVirtClient()
+				Expect(err).ToNot(HaveOccurred())
+				kv, err := c.KubevirtV1().KubeVirts(v1.NamespaceAll).List(context.Background(), v1.ListOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				Expect(kv.Items).To(HaveLen(1))
+				kubevirtNS = kv.Items[0].Namespace
+			})
```
Comment (enp0s3) [inline tests/kubevirt_api_lifecycle_automation_test.go]: @Barakmor1 I would suggest to do it before each test in order to make sure the cluster is healthy for running the e2e

File: tests/kubevirt_api_lifecycle_automation_test.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+	vm := libvmi.NewVirtualMachine(
+		vmi,
+		libvmi.WithRunStrategy(virt1.RunStrategyAlways),
+	)
+	vm, err = c.KubevirtV1().VirtualMachines(f.Namespace.Name).Create(context.Background(), vm, v1.CreateOptions{})
+	Expect(err).ToNot(HaveOccurred())
+	err = c.KubevirtV1().VirtualMachines(f.Namespace.Name).Delete(context.Background(), vm.Name, v1.DeleteOptions{})
+	Expect(err).ToNot(HaveOccurred())
+
+	return vm.Spec.Template.Spec.Domain.Machine.Type
```
Comment (enp0s3) [inline tests/kubevirt_api_lifecycle_automation_test.go]: @Barakmor1 You can use dry-run here. Mutating admitters should be applied in dry run.

File: tests/kubevirt_api_lifecycle_automation_test.go [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func activateKubevirtApiLifecycleAutomationJob(f *framework.Framework, machineTypeGlob, restartRequired string) {
+	newJob := operator.CreateKubevirtApiLifecycleAutomation(f.KubevirtApiLifecycleAutomationNamespace, machineTypeGlob, v1.NamespaceAll, restartRequired, "", "", f.KubevirtApiLifecycleAutomationImage, "Always")
```
Comment (enp0s3) [inline tests/kubevirt_api_lifecycle_automation_test.go:162]: For follow-up: try to avoid using product code.

**Comment by enp0s3** [comment]
@Barakmor1 I would squash the last commit with the first one. Other than that I think we can merge it and iterate over improvements in follow ups.

File: tests/kubevirt_api_lifecycle_automation_test.go (hunk 11 above)
Comment (Barakmor1) [reply tests/kubevirt_api_lifecycle_automation_test.go]: Yes, thanks for the great advice :)

**Comment by Barakmor1** [comment]
Hey @enp0s3,

Thanks for the review. I’ve addressed all the comments. Please let me know if I missed anything.

**Pull request merged**